"panic: runtime error: slice bounds out of range" when create session

Env is Cassandra 3.5, standalone, gocql is latest master build.

>[root@pvg-ta-super test]# cat test_gocql.go 
package main

>import (
	"github.com/gocql/gocql"
)

>func main() {
	// connect to the cluster
	cluster := gocql.NewCluster("super")
	cluster.Keyspace = "demo"
	session, _ := cluster.CreateSession()
        defer session.Close()
}`

Panic trace
>[root@super test]# go run test_gocql.go 
panic: runtime error: slice bounds out of range

>goroutine 24 [running]:
panic(0x6f2020, 0xc82000a090)
	/root/ken/go/src/runtime/panic.go:481 +0x3e6
github.com/gocql/gocql.readHeader(0x7f5a6e9d5488, 0xc82007c4e0, 0xc820080c78, 0x8, 0x8, 0x0, 0x0, 0xffffffffffffff00, 0x0, 0x0, ...)
	/root/ken/gocode/src/github.com/gocql/gocql/frame.go:358 +0x773
github.com/gocql/gocql.(*Conn).recv(0xc8200d4000, 0x0, 0x0)
	/root/ken/gocode/src/github.com/gocql/gocql/conn.go:448 +0x118
github.com/gocql/gocql.Connect.func1(0xc82007c600, 0xc8200d4000, 0xc82007c660, 0x7f5a6e9d6720, 0xc820083b00)
	/root/ken/gocode/src/github.com/gocql/gocql/conn.go:220 +0x71
created by github.com/gocql/gocql.Connect
	/root/ken/gocode/src/github.com/gocql/gocql/conn.go:230 +0x8d5
exit status 2`

--- a/conn.go
+++ b/conn.go
@@ -127,7 +127,7 @@ type Conn struct {
 	timeout time.Duration
 	cfg     *ConnConfig
 
-	headerBuf []byte
+	headerBuf [maxFrameHeaderSize]byte
 
 	streams *streams.IDGenerator
 	mu      sync.RWMutex
@@ -175,11 +175,6 @@ func Connect(host *HostInfo, addr string, cfg *ConnConfig,
 		return nil, err
 	}
 
-	headerSize := 8
-	if cfg.ProtoVersion > protoVersion2 {
-		headerSize = 9
-	}
-
 	c := &Conn{
 		conn:         conn,
 		r:            bufio.NewReader(conn),
@@ -191,7 +186,6 @@ func Connect(host *HostInfo, addr string, cfg *ConnConfig,
 		errorHandler: errorHandler,
 		compressor:   cfg.Compressor,
 		auth:         cfg.Authenticator,
-		headerBuf:    make([]byte, headerSize),
 		quit:         make(chan struct{}),
 		session:      session,
 		streams:      streams.New(cfg.ProtoVersion),
@@ -445,7 +439,7 @@ func (c *Conn) recv() error {
 	}
 
 	// were just reading headers over and over and copy bodies
-	head, err := readHeader(c.r, c.headerBuf)
+	head, err := readHeader(c.r, c.headerBuf[:])
 	if err != nil {
 		return err
 	}

--- a/frame.go
+++ b/frame.go
@@ -231,6 +231,8 @@ var (
 	ErrFrameTooBig = errors.New("frame length is bigger than the maximum allowed")
 )
 
+const maxFrameHeaderSize = 9
+
 func writeInt(p []byte, n int32) {
 	p[0] = byte(n >> 24)
 	p[1] = byte(n >> 16)
@@ -366,7 +368,7 @@ func readHeader(r io.Reader, p []byte) (head frameHeader, err error) {
 	head.flags = p[1]
 
 	if version > protoVersion2 {
-		if len(p) < 9 {
+		if len(p) != 9 {
 			return frameHeader{}, fmt.Errorf("not enough bytes to read header require 9 got: %d", len(p))
 		}
 
@@ -374,7 +376,7 @@ func readHeader(r io.Reader, p []byte) (head frameHeader, err error) {
 		head.op = frameOp(p[4])
 		head.length = int(readInt(p[5:]))
 	} else {
-		if len(p) < 8 {
+		if len(p) != 8 {
 			return frameHeader{}, fmt.Errorf("not enough bytes to read header require 8 got: %d", len(p))
 		}
 

